azure-quantum/azure/quantum/workspace.py

```diff
@@ -132,110 +106,91 @@ def __init__(
-                # -> match.group(3): The Azure Quantum workspace name
-                subscription_id = match.group(1)
-                resource_group = match.group(2)
-                name = match.group(3)
-
-        if not subscription_id or not resource_group or not name:
-            raise ValueError(
-                "Azure Quantum workspace not fully specified."
-                + "Please specify either a valid resource ID "
-                + "or a valid combination of subscription ID,"
-                + "resource group name, and workspace name."
-            )
+        connection_params = WorkspaceConnectionParams(
+            location=location,
+            subscription_id=subscription_id,
+            resource_group=resource_group,
+            workspace_name=name,
+            credential=credential,
+            resource_id=resource_id,
+            user_agent=user_agent,
+            **kwargs
+        ).default_from_env_vars()
+        logger.info("Using %s environment.", connection_params.environment)
+
+        connection_params.assert_complete()
 
-        if not location:
-            raise ValueError(
-                "Azure Quantum workspace does not have an associated location. " +
-                "Please specify the location associated with your workspace.")
-
-        # Temporarily using a custom _DefaultAzureCredential
-        # instead of Azure.Identity.DefaultAzureCredential
-        # See _DefaultAzureCredential documentation for more info.
-        if credential is None:
-            credential = _DefaultAzureCredential(exclude_interactive_browser_credential=False,
-                                                 exclude_shared_token_cache_credential=True,
-                                                 subscription_id=subscription_id,
-                                                 arm_base_url=ARM_BASE_URL)
-
-        self.credentials = credential
-        self.name = name
-        self.resource_group = resource_group
-        self.subscription_id = subscription_id
-        self.storage = storage
-        self._user_agent = user_agent
-
-        # Convert user-provided location into names
-        # recognized by Azure resource manager.
-        # For example, a customer-provided value of
-        # "West US" should be converted to "westus".
-        self.location = "".join(location.split()).lower()
+        connection_params.on_new_client_request = self.on_new_client_request
+
+        self._connection_params = connection_params
+        self._storage = storage
 
         # Create QuantumClient
         self._client = self._create_client()
 
-    def _create_client(self) -> QuantumClient:
-        endpoint = BASE_URL(self.location)
-        logger.debug(
-            f"Creating client for: subs:{self.subscription_id},"
-            + f"rg={self.resource_group}, ws={self.name}, frontdoor={endpoint}"
-        )
+    def on_new_client_request(self):
```

Comment:
Why we need to create_client() to _client twice?

Reply:
We need to recreate the client when the user_agent changes.
This happens in the cirq and qiskit clients.